// Explorer/Explorer.js
function Explorer(fileOpener, filesLister, bus, getFilename, localStorage) {
    var self = this;

    this.selectFile = function (file, app) {
        bus("FileSelected").broadcast(file);
    }

    this.refreshFolder = function (filesLister, folder) {
        var newContent = filesLister(folder.path);
        folder.files = newContent.files;
        folder.folders = newContent.folders;
    }

    this.selectFolder = function (folder) {
        folder.expanded = !folder.expanded
        if (folder.expanded && folder.files.length == 0 && folder.folders.length == 0) self.refreshFolder(filesLister, folder);
    }

    this.setTree = function(tree){
        this.tree = tree;
    }
    
    this.addFile = function(file){
        this.folders.push({
            name:getFilename(file),
            expanded:false,
            path:file+"\\",
            files:[],
            folders:[]
        });
    }


    var store = localStorage["Explorer"];
    if(store)store = JSON.parse(store);
    this.folders = store?store:[];
    
    bus("ApplicationClosing").subscribe(function(){
        localStorage["Explorer"] = JSON.stringify(self.folders);
    });
}

module.exports = Explorer;
